migrator/migrator.go

```diff
@@ -0,0 +1,269 @@
+		return nil, microerror.Maskf(invalidConfigError, fmt.Sprintf("%T.DockerRegistry must not be empty", config))
+	}
+	if config.EtcdCaFile == "" {
+		return nil, microerror.Maskf(invalidConfigError, fmt.Sprintf("%T.EtcdCaFile must not be empty", config))
+	}
+	if config.EtcdCertFile == "" {
+		return nil, microerror.Maskf(invalidConfigError, fmt.Sprintf("%T.EtcdCertFile must not be empty", config))
+	}
+	if config.EtcdEndpoint == "" {
+		return nil, microerror.Maskf(invalidConfigError, fmt.Sprintf("%T.EtcdEndpoint must not be empty", config))
+	}
+	if config.EtcdKeyFile == "" {
+		return nil, microerror.Maskf(invalidConfigError, fmt.Sprintf("%T.EtcdKeyFile must not be empty", config))
+	}
+
+	etcdClient, err := createEtcdClient(config.EtcdCaFile, config.EtcdCertFile, config.EtcdKeyFile, config.EtcdEndpoint)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sClient, err := createK8SClient()
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	m := &Migrator{
+		baseDomain:        config.BaseDomain,
+		dockerRegistry:    config.DockerRegistry,
+		etcdStartingIndex: config.EtcdStartingIndex,
+		masterNodeLabel:   config.MasterNodeLabel,
+
+		etcdClient: etcdClient,
+		k8sClient:  k8sClient,
+	}
+
+	return m, nil
+}
+
+func (m *Migrator) Run() error {
+	defer m.etcdClient.Close()
+	ctx := context.Background()
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, time.Second*20)
+	defer cancel()
+
+	nodeNames, err := getMasterNodes(m.k8sClient, m.masterNodeLabel)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	memberListResponse, err := m.etcdClient.Cluster.MemberList(ctxWithTimeout)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	memberCount := len(memberListResponse.Members)
+
+	fmt.Printf("Found %d etcd members in the cluster.\n", memberCount)
+
+	if memberCount == masterNodeCount {
+		fmt.Printf("Etcd cluster already has 3 nodes. Nothing to do. Exiting.\n")
+		return nil
+	} else if memberCount == 2 {
+		// continue migration that was interrupted in the middle of the process
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 3)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+	} else if memberCount == 1 {
+		//  ensure that first node has proper etcd peer url set to etcd1.xxxx.xxxx.xxx
+		err = m.fixFirstNodePeerUrl(ctx, memberListResponse.Members)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		// add second node to the etcd cluster
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 2)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// add third node to the etcd cluster
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 3)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	} else {
+		fmt.Printf("unexpected number of nodes in etcd cluster\n")
+		return microerror.Maskf(executionFailedError, fmt.Sprintf("found %d nodes in etcd cluster", memberCount))
+	}
+
+	fmt.Printf("ETCD cluster migration succesfuly finished.\n\n")
+	fmt.Printf("Sleeping forever.\n")
```

Comment:
Why sleeping forever and not just shutting down?

Reply:
As now this is deployment, and pod under deployment that is finished will be restarted by the k8s controller.

But that got me thinking maybe we can run this as a job so once it's finished, it won't start again.

Reply:
Running it as a job once makes sense to me when it only should act once.

Reply:
Yep, already migrated to a job and remove the sleeping forever